File: dp/tle.cpp
#include <chrono>
#include <cstdlib>
#include <filesystem>
#include <iostream>
#include <optional>
#include <string>

namespace fs = std::filesystem;

enum class GeneratorMode {
	Cpp,
	Python,
};

struct Options {
	GeneratorMode generator_mode = GeneratorMode::Cpp;
	std::optional<double> time_limit_ms;
};

std::string quote(const std::string& value) {
	return "\"" + value + "\"";
}

bool has_required_files(const fs::path& dir) {
	return fs::exists(dir / "solve.cpp") && (fs::exists(dir / "data.cpp") || fs::exists(dir / "data.py"));
}

std::optional<fs::path> locate_base_dir(const fs::path& start) {
	fs::path current = start;
	for (int depth = 0; depth < 6; ++depth) {
		if (has_required_files(current)) {
			return current;
		}
		if (has_required_files(current / "dp")) {
			return current / "dp";
		}
		if (!current.has_parent_path()) {
			break;
		}
		fs::path parent = current.parent_path();
		if (parent == current) {
			break;
		}
		current = parent;
	}
	return std::nullopt;
}

int run(const std::string& command) {
	return std::system(command.c_str());
}

std::string find_compiler() {
	for (const std::string& candidate : {"g++-15", "g++-14", "g++-13", "g++-12", "g++-11", "g++-10", "g++", "clang++", "c++"}) {
		if (run("command -v " + candidate + " >/dev/null 2>&1") == 0) {
			return candidate;
		}
	}
	return "";
}

std::optional<Options> parse_options(int argc, char* argv[]) {
	Options options;
	for (int i = 1; i < argc; ++i) {
		std::string arg = argv[i];
		if (arg == "cpp") {
			options.generator_mode = GeneratorMode::Cpp;
			continue;
		}
		if (arg == "py" || arg == "python") {
			options.generator_mode = GeneratorMode::Python;
			continue;
		}
		try {
			size_t pos = 0;
			double value = std::stod(arg, &pos);
			if (pos != arg.size() || value < 0) {
				return std::nullopt;
			}
			options.time_limit_ms = value;
		} catch (...) {
			return std::nullopt;
		}
	}
	return options;
}

int main(int argc, char* argv[]) {
	fs::path exe_dir = fs::absolute(argv[0]).parent_path();
	std::optional<fs::path> base_dir = locate_base_dir(fs::current_path());
	if (!base_dir) {
		base_dir = locate_base_dir(exe_dir);
	}
	if (!base_dir) {
		std::cerr << "Failed to locate dp directory.\n";
		return 1;
	}
	fs::path include_dir = base_dir->parent_path() / ".vscode" / "include";
	fs::path artifact_dir = *base_dir / ".tmp";
	std::error_code ec;
	fs::current_path(*base_dir, ec);
	if (ec) {
		std::cerr << "Failed to enter dp directory: " << ec.message() << '\n';
		return 1;
	}

	std::optional<Options> options = parse_options(argc, argv);
	if (!options) {
		std::cerr << "Usage: " << argv[0] << " [cpp|py] [time_limit_ms]\n";
		return 2;
	}

	std::string compiler = find_compiler();
	if (compiler.empty()) {
		std::cerr << "No C++ compiler found.\n";
		return 1;
	}
	fs::create_directories(artifact_dir);

	std::string compile_flags = " -std=c++23 -O2 -I " + quote(include_dir.string());
	fs::path solve_bin = artifact_dir / "solve";
	fs::path data_bin = artifact_dir / "data";
	fs::path data_in = artifact_dir / "data.in";
	fs::path solve_out = artifact_dir / "solve.out";
	if (run(compiler + compile_flags + " solve.cpp -o " + quote(solve_bin.string()))) return 1;
	if (options->generator_mode == GeneratorMode::Cpp && run(compiler + compile_flags + " data.cpp -o " + quote(data_bin.string()))) return 1;

	if (options->generator_mode == GeneratorMode::Cpp) {
		if (run(quote(data_bin.string()) + " > " + quote(data_in.string()))) return 1;
	} else {
		if (run("python3 data.py > " + quote(data_in.string()))) return 1;
	}

	auto start = std::chrono::steady_clock::now();
	int code = run(quote(solve_bin.string()) + " < " + quote(data_in.string()) + " > " + quote(solve_out.string()));
	auto finish = std::chrono::steady_clock::now();
	double elapsed_ms = std::chrono::duration<double, std::milli>(finish - start).count();

	std::cout << "Elapsed: " << elapsed_ms << " ms\n";
	if (code != 0) {
		std::cerr << "solve exited with code " << code << '\n';
		return 1;
	}
	if (options->time_limit_ms && elapsed_ms > *options->time_limit_ms) {
		std::cout << "Potential TLE\n";
		return 3;
	}
	return 0;
}

File: dp/tle.py
from pathlib import Path
import shutil
import subprocess
import sys
import time

BASE_DIR = Path(__file__).resolve().parent
INCLUDE_DIR = BASE_DIR.parent / ".vscode" / "include"
ARTIFACT_DIR = BASE_DIR / ".tmp"


def find_compiler() -> str:
    for candidate in ("g++-15", "g++-14", "g++-13", "g++-12", "g++-11", "g++-10", "g++", "clang++", "c++"):
        if shutil.which(candidate):
            return candidate
    raise FileNotFoundError("No C++ compiler found")


def parse_generator_mode() -> str:
    generator_mode = "cpp"
    time_limit_ms = None
    for arg in sys.argv[1:]:
        mode = arg.lower()
        if mode in {"cpp", "py", "python"}:
            generator_mode = "py" if mode in {"py", "python"} else "cpp"
            continue
        try:
            value = float(arg)
        except ValueError as exc:
            raise SystemExit(f"Usage: {sys.argv[0]} [cpp|py] [time_limit_ms]") from exc
        if value < 0:
            raise SystemExit(f"Usage: {sys.argv[0]} [cpp|py] [time_limit_ms]")
        time_limit_ms = value
    return generator_mode, time_limit_ms


def main() -> int:
    generator_mode, time_limit_ms = parse_generator_mode()
    compiler = find_compiler()
    ARTIFACT_DIR.mkdir(exist_ok=True)
    subprocess.run(
        [compiler, "-std=c++23", "-O2", "-I", str(INCLUDE_DIR), "solve.cpp", "-o", str(ARTIFACT_DIR / "solve")],
        cwd=BASE_DIR,
        check=True,
    )
    data_in = ARTIFACT_DIR / "data.in"
    solve_out = ARTIFACT_DIR / "solve.out"
    with data_in.open("w", encoding="utf-8") as data_in_file:
        if generator_mode == "cpp":
            subprocess.run(
                [compiler, "-std=c++23", "-O2", "-I", str(INCLUDE_DIR), "data.cpp", "-o", str(ARTIFACT_DIR / "data")],
                cwd=BASE_DIR,
                check=True,
            )
            subprocess.run([str(ARTIFACT_DIR / "data")], cwd=BASE_DIR, stdout=data_in_file, check=True)
        else:
            subprocess.run([sys.executable, "data.py"], cwd=BASE_DIR, stdout=data_in_file, check=True)
    with data_in.open("r", encoding="utf-8") as data_in_file, solve_out.open("w", encoding="utf-8") as solve_out_file:
        start = time.perf_counter()
        subprocess.run([str(ARTIFACT_DIR / "solve")], cwd=BASE_DIR, stdin=data_in_file, stdout=solve_out_file, check=True)
        elapsed_ms = (time.perf_counter() - start) * 1000
    print(f"Elapsed: {elapsed_ms:.3f} ms")
    if time_limit_ms is not None and elapsed_ms > time_limit_ms:
        print("Potential TLE")
        return 3
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
